tools/build-flow-types.js

```diff
@@ -0,0 +1,155 @@
+ */
+const getExportsFrom = async (
+    filePath /*: string*/,
+) /*: Promise<Array<string>>*/ => {
+    /**
+     * We assume the file we're given is an export as that's how our
+     * additional-exports syntax works.
+     */
+    const files = [path.basename(filePath)];
+    try {
+        const fileContent = await readFileAsync(filePath);
+
+        /**
+         * First regex gets the block of file names as indicated by
+         * the @additional-exports marker, between it's [] brackets.
+         * We don't need to loop the regex here as we only want the first
+         * match. We don't care if there are more.
+         *
+         * The list should have one file per line within double quotes
+         * and ending in a comma (went with the comma to make it feel like
+         * our coding standard).
+         */
+        const additionalFilesMatch = regexAdditionalFiles.exec(fileContent);
+        if (additionalFilesMatch != null && additionalFilesMatch[1] != null) {
+            /**
+             * We have a match, so let's now parse each file and add that to
+             * our list.
+             */
+            let additionalFileMatch;
+            do {
+                additionalFileMatch = regexExtractFile.exec(
+                    additionalFilesMatch[1],
+                );
+                if (
+                    additionalFileMatch != null &&
+                    additionalFileMatch[1] != null
+                ) {
+                    files.push(additionalFileMatch[1]);
+                }
+            } while (additionalFileMatch != null);
+        }
+        /**
+         * We will worry about duplicates later.
+         */
+        return files;
+    } catch (e) {
+        throw new Error(`Unable to process file: ${filePath} ${e}`);
+    }
+};
+
+/**
+ * Write out a .js.flow file in our disk for the file specified.
+ */
+const writeFlowFile = (
+    exportFromPath /*: string*/,
+    outputPath /*: string*/,
+) /*: Promise<void>*/ => {
+    const contents = ["// @flow", `export * from "${exportFromPath}";`].join(
+        "\n",
+    );
+    return writeFileAsync(outputPath, contents, "utf8");
+};
+
+/**
+ * Output flow type files to our dist for our dist packages.
+ */
+const outputFlowTypesForDist = async () /*: Promise<void>*/ => {
+    const rootDir = ancesdir(__dirname);
+    const srcDir = path.normalize(path.join(rootDir, "src"));
+    const distDir = path.normalize(path.join(rootDir, "dist"));
+
+    /**
+     * We only care about packages in our distribution.
+     */
+    const packages = fs.readdirSync(distDir);
+    console.log(`Building flow files for ${packages.join(", ")}`);
+    for (const pkg of packages) {
+        const pkgSrcDir = path.join(srcDir, pkg);
+        const pkgDistDir = path.join(distDir, pkg);
+
+        const srcIndexJSPath = path.join(pkgSrcDir, "index.js");
+        const filesThatNeedTypes = await getExportsFrom(srcIndexJSPath);
+        for (const file of filesThatNeedTypes) {
```

Comment:
This doesn't handle `index.js` files.  Are those handled manually at the moment?

Reply:
It does. See the `getExportsFrom` method

Reply:
:+1: I missed the `const files = [path.basename(filePath)];`.